mitosheet/mitosheet/parser.py

```diff
@@ -256,69 +420,110 @@ def find_column_headers(match: Any) -> Any:
+        formula_label: Union[str, bool, int, float],
         string_matches: List,
+        df: pd.DataFrame,
         df_name: str
     ) -> Tuple[str, Set[ColumnHeader]]:
     """
-    Returns a modified formula, where the column headers in the string
-    have been replaced with references to the dataframe.
+    Returns a modified formula, where:
+    1. column headers have been replaced with references to the dataframe
+    2. index labels that follow those column headers are entirely removed
 
-    We except that they are _not_ contained within any reference (e.g.
-    not within quotes, or anything else for that matter).
+    If the index label referenced after a column header is not the the same as the
+    formula_label, then we change the column header to have a .shift that shifts
+    it the same relative amount. 
     """
     # We get all the matches first, from end to start order
-    column_header_match_tuples = get_column_header_match_tuples(
+    parser_matches = get_column_header_and_index_matches(
         formula,
-        column_headers,
-        string_matches
+        formula_label,
+        string_matches,
+        df
     )
+
+    column_headers = set()
     
     # Then, go through from the end to the start, and actually replace all the column headers
-    for column_header, match in column_header_match_tuples:
-        start = match.start()
-        end = match.end()
-
-        if isinstance(column_header, str):
-            replace_string = f'{df_name}[\'{column_header}\']'
-        elif isinstance(column_header, datetime.datetime):
-            replace_string = f'{df_name}[pd.to_datetime(\'{column_header}\')]'
-        elif isinstance(column_header, datetime.timedelta):
-            replace_string = f'{df_name}[pd.to_timedelta(\'{column_header}\')]'
-        else:
-            replace_string = f'{df_name}[{column_header}]'
-
-        formula = formula[:start] + replace_string + formula[end:]
-
-    return formula, set([ch for (ch, _) in column_header_match_tuples])
+    # and remove all of the index labels
+    for match in parser_matches:
+        match_type = match['type']
+        start = match['match_range'][0]
+        end = match['match_range'][1]
+
+        if match_type == COLUMN_HEADER_MATCH_TYPE:
+            column_header = match['parsed']
+            column_headers.add(column_header)
+
+            if isinstance(column_header, str):
+                replace_string = f'{df_name}[\'{column_header}\']'
+            elif isinstance(column_header, datetime.datetime):
+                replace_string = f'{df_name}[pd.to_datetime(\'{column_header}\')]'
+            elif isinstance(column_header, datetime.timedelta):
+                replace_string = f'{df_name}[pd.to_timedelta(\'{column_header}\')]'
+            else:
+                replace_string = f'{df_name}[{column_header}]'
+
+            # If there is a row offset, then we use this to put the column header in an offset function
```

Comment:
"put the column header in an offset function" -> "shift the series the correct number of rows".